pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java

```diff
@@ -644,6 +646,25 @@ private CompletableFuture<Optional<CandidateBrokerResult>> getCandidateBrokerRes
 
     }
 
+    private CompletableFuture<Optional<LeaderBroker>> getOrWaitForLeader(AtomicInteger retries) {
+        return pulsar.getLeaderElectionService().readCurrentLeader()
+                .handle((leaderBroker, t) -> {
+                    CompletableFuture<Optional<LeaderBroker>> retval;
+                    if ((t != null || !leaderBroker.isPresent()) && retries.getAndIncrement() < 3) {
+                        // retry after a delay of 5 seconds
+                        retval =
+                                CompletableFuture.supplyAsync(() -> null,
+                                                CompletableFuture.delayedExecutor(5, SECONDS))
```

Comment:
which thread pool will run this execution ? (I guess it would be the system common pool)
should we use a well-known thread pool ?

Reply:
it's not really a problem to use the system's pool for delayed executor. However, delayed executor isn't available in Java 8 and using this will make backporting harder. I'll use Pulsar's existing scheduler for this purpose so that backporting will be easier.